### PR TITLE
feat: create api on oracle to fetch spz information given a spu

### LINKED ISSUE #795: Create API on Oracle to fetch SPZ information given a SPU ID
Task:

Create GET endpoint on `Oracle` to receive a SPU ID and respond with the SPZ information. Build for the condition that parent trees are tested parent trees. 

Here's the queries you can use:
```SQL
-- Get the tested parent tree area of use id: E.g.: '40'
SELECT * FROM the.tested_pt_area_of_use tested WHERE tested.SEED_PLAN_UNIT_ID  = 7; -- 40

-- Then make sure the relationship between **tested_parent_tree** and **seed_plan_unit** exists: 
SELECT * FROM the.tested_pt_area_of_use_spu aspu WHERE aspu.TESTED_PT_AREA_OF_USE_ID = 40 AND aspu.SEED_PLAN_UNIT_ID = 7; -- should exist. If not, throw error!!

-- Then get the seed plan zone id, elevation_min, and elevation_max e.g.: 1284, 1 and 700
SELECT * FROM the.seed_plan_unit spu WHERE spu.SEED_PLAN_UNIT_ID = 7;

-- Then get the seed plan zone information
SELECT * FROM the.seed_plan_zone spz WHERE spz.SEED_PLAN_ZONE_ID = 1284;
```

### Acceptance criteria

- API returning: 
  - `seed_plan_zone_id` E.g.: `1284`
  - `genetic_class_code` E.g.: `'A'`
  - `seed_plan_zone_code` E.g.: `'M'`
  - `vegetation_code` E.g.: `FDC`
  - `elevation_min` E.g.: `1`
  - `elevation_max` E.g.: `700`
  - IMPORTANT: the API should be able to receive ONE or MORE spu ids.

### Position
- Subtask 2 of 4 of https://github.com/bcgov/nr-spar/issues/788

### Dependencies
- None

More info:
- Write up for SPZ look up rules for tested parent trees located in the bottom half of the document here: https://apps.nrs.gov.bc.ca/int/confluence/display/FSADT2/Defining+Seedlot+Geographic+data
- Subtask of https://github.com/bcgov/nr-spar/issues/788

### PR DESCRIPTION
**Pull request by RMCampos**
# Description
Closes #795 

#### New
- Methods for fetching SPZ (Seed Plan Zone) information given a list of SPU (Seed Planning Unit) id list from **Oracle THE** database;
- Entity and Repository classes required for gathering the information.

#### Changed
- `OrchardEndpoint` class
- `OrchardService` class

#### Removed
- None;

### How was this tested?
- [ ] 🧠 Not needed
- [x] 👀 Eyeball
- [x] 🤖 Added tests

No post-deployment tasks required.

###  What gif/image best describes this PR or how it makes you feel?

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Frontend](https://nr-spar-10-frontend.apps.silver.devops.gov.bc.ca/)
[Backend](https://nr-spar-10-backend.apps.silver.devops.gov.bc.ca/actuator/health)
[Oracle-API](https://nr-spar-10-oracle-api.apps.silver.devops.gov.bc.ca/actuator/health)

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/nr-spar/actions/workflows/merge-main.yml)

### CODE REVIEW

**Review by RMCampos — COMMENTED** [review]
Great job! 💯 Only some comments and quick fixes.

File: oracle-api/src/main/java/ca/bc/gov/backendstartapi/endpoint/OrchardEndpoint.java [hunk 1]
```diff
@@ -184,4 +186,28 @@ public ResponseEntity<List<SameSpeciesTreeDto>> findParentTreesWithVegCode(
...
+      responses = {
+        @ApiResponse(
+            responseCode = "200",
+            description = "List with found records or an empty list"),
+        @ApiResponse(
+            responseCode = "401",
+            description = "Access token is missing or invalid",
+            content = @Content(schema = @Schema(implementation = Void.class)))
+      })
+  public List<SpzBySpuDto> getSpzInformation(
```
[the remark this answers is not in the document]
Comment (RMCampos) [reply oracle-api/src/main/java/ca/bc/gov/backendstartapi/endpoint/OrchardEndpoint.java:209]: Changed to `SeedPlanZoneDto`

File: oracle-api/src/main/java/ca/bc/gov/backendstartapi/service/OrchardService.java [hunk 2]
```diff
@@ -308,4 +334,74 @@ private ParentTreeGeneticInfoDto convertToParentTreeGenInfoDto(ParentTreeEntity
...
+  /**
+   * Get SPZ information given a list of SPU Ids.
+   *
+   * @param spuIds A list of SPU ID to be fetched.
+   * @return A List of {@link SpzBySpuDto}
+   */
+  public List<SpzBySpuDto> getSpzInformationBySpu(List<Integer> spuIds) {
+    SparLog.info("Getting SPZ information for SPU IDs {}", spuIds);
+
+    // Finds all TESTES PT AREA OF USE and map them for spu
```
[the remark this answers is not in the document]
Comment (RMCampos) [reply oracle-api/src/main/java/ca/bc/gov/backendstartapi/service/OrchardService.java:347]: Nope. Removed

File: oracle-api/src/main/java/ca/bc/gov/backendstartapi/service/OrchardService.java [hunk 3]
```diff
@@ -308,4 +334,74 @@ private ParentTreeGeneticInfoDto convertToParentTreeGenInfoDto(ParentTreeEntity
...
+    List<TestedPtAreaOfUse> testedList =
+        testedPtAreaofUseRepository.findAllBySeedPlanUnitIdIn(spuIds);
+
+    if (testedList.isEmpty()) {
+      SparLog.info("No testes parent tree area of use found!");
+      return List.of();
+    }
+
+    List<SpzBySpuDto> responseDtpList = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (RMCampos) [reply oracle-api/src/main/java/ca/bc/gov/backendstartapi/service/OrchardService.java:356]: Done

File: oracle-api/src/main/java/ca/bc/gov/backendstartapi/endpoint/OrchardEndpoint.java [hunk 4]
```diff
@@ -184,4 +186,28 @@ public ResponseEntity<List<SameSpeciesTreeDto>> findParentTreesWithVegCode(
...
     }
   }
+
+  /**
+   * Gets all SPZ information given a list of SPU IDs.
+   *
+   * @param spuIds A list of SPU ID to be fetched.
+   * @return A list of {@link SpzBySpuDto} containing the results or an empty list.
+   */
+  @GetMapping(path = "/spz-information/{spuIds}", produces = MediaType.APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (RMCampos) [reply oracle-api/src/main/java/ca/bc/gov/backendstartapi/endpoint/OrchardEndpoint.java]: Done https://github.com/bcgov/nr-spar/pull/860/commits/f00b60c99831584b1404ee0b81d0a3608bd07a4a

File: oracle-api/src/main/java/ca/bc/gov/backendstartapi/endpoint/OrchardEndpoint.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (RMCampos) [reply oracle-api/src/main/java/ca/bc/gov/backendstartapi/endpoint/OrchardEndpoint.java]: Done https://github.com/bcgov/nr-spar/pull/860/commits/f00b60c99831584b1404ee0b81d0a3608bd07a4a

File: oracle-api/src/main/java/ca/bc/gov/backendstartapi/service/OrchardService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (RMCampos) [reply oracle-api/src/main/java/ca/bc/gov/backendstartapi/service/OrchardService.java]: Done https://github.com/bcgov/nr-spar/pull/860/commits/f00b60c99831584b1404ee0b81d0a3608bd07a4a

**Review by DerekRoberts — APPROVED** [review]
Nice.  :)

File: oracle-api/src/main/java/ca/bc/gov/backendstartapi/dto/SeedPlanZoneDto.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+  @Schema(description = "A code describing various Seed Planning Zones.", example = "M")
+  private String seedPlanZoneCode;
+
+  @Schema(description = "A code describing various Vegetation Species.", example = "FDC")
+  private String vegetationCode;
+
+  @Schema(
+      description =
+          "Minimum elevation (metres) for a specific elevation band for the Seed Planning Unit",
```
Comment (mgaseta) [inline oracle-api/src/main/java/ca/bc/gov/backendstartapi/dto/SeedPlanZoneDto.java:34]: Non-blocking: Typo here `metres` should be `meters`

File: oracle-api/src/test/resources/application.properties [hunk 6]
```diff
@@ -6,7 +6,7 @@ logging.level.root = INFO
 
 # Database, datasource and JPA
 spring.datasource.driver-class-name=org.h2.Driver
-spring.datasource.url = jdbc:h2:~/testdb;DB_CLOSE_ON_EXIT=FALSE
+spring.datasource.url = jdbc:h2:mem:oracledb;DB_CLOSE_ON_EXIT=TRUE
```
Comment (mgaseta) [inline oracle-api/src/test/resources/application.properties:9]: Question, non-blocking: what does this change means?

**Review by mgaseta — APPROVED** [review]
Nice job! SPZ FTW! :100: :1st_place_medal: 
Added only a fez non-blocking mark-ups :)

File: oracle-api/src/test/resources/application.properties (hunk 6 above)
Comment (RMCampos) [reply oracle-api/src/test/resources/application.properties:9]: That's how we make the tests running faster.

File: oracle-api/src/main/java/ca/bc/gov/backendstartapi/dto/SeedPlanZoneDto.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
+package ca.bc.gov.backendstartapi.dto;
+
+import io.swagger.v3.oas.annotations.media.Schema;
+import lombok.Getter;
+import lombok.Setter;
+
+/** This class represents a JSON response when requesting SPZ information from a SPU. */
+@Getter
+@Setter
+@Schema(description = "represents a JSON response when requesting SPZ information from a SPU.")
```
[the remark this answers is not in the document]
Comment (RMCampos) [reply oracle-api/src/main/java/ca/bc/gov/backendstartapi/dto/SeedPlanZoneDto.java]: Done! Thanks https://github.com/bcgov/nr-spar/pull/860/commits/ed9039879e726a6f7993912c49e2f0bdbc0a4912

File: oracle-api/src/test/resources/application.properties (hunk 6 above)
Comment (mgaseta) [reply oracle-api/src/test/resources/application.properties:9]: Nice!

**Pull request merged**